rs/rosetta-api/icrc1/tests/golden_state_upgrade_downgrade.rs

```diff
@@ -0,0 +1,146 @@
+const ICPANDA: (&str, &str) = ("druyg-tyaaa-aaaaq-aactq-cai", "ICPanda");
+const ICPCC: (&str, &str) = ("lrtnw-paaaa-aaaaq-aadfa-cai", "ICPCC");
+const ICX: (&str, &str) = ("rffwt-piaaa-aaaaq-aabqq-cai", "ICX");
+const KINIC: (&str, &str) = ("73mez-iiaaa-aaaaq-aaasq-cai", "Kinic");
+const DECIDEAI: (&str, &str) = ("xsi2v-cyaaa-aaaaq-aabfq-cai", "DecideAI");
+const MOTOKO: (&str, &str) = ("k45jy-aiaaa-aaaaq-aadcq-cai", "Motoko");
+const NEUTRINITE: (&str, &str) = ("f54if-eqaaa-aaaaq-aacea-cai", "Neutrinite");
+const NUANCE: (&str, &str) = ("rxdbk-dyaaa-aaaaq-aabtq-cai", "Nuance");
+const OPENFPL: (&str, &str) = ("ddsp7-7iaaa-aaaaq-aacqq-cai", "OpenFPL");
+const ORIGYN: (&str, &str) = ("lkwrt-vyaaa-aaaaq-aadhq-cai", "Origyn");
+const SNEED: (&str, &str) = ("hvgxa-wqaaa-aaaaq-aacia-cai", "Sneed");
+const SONIC: (&str, &str) = ("qbizb-wiaaa-aaaaq-aabwq-cai", "Sonic");
+const TRAX: (&str, &str) = ("emww2-4yaaa-aaaaq-aacbq-cai", "Trax");
+const WATERNEURON: (&str, &str) = ("jcmow-hyaaa-aaaaq-aadlq-cai", "WaterNeuron");
+const YUKU: (&str, &str) = ("atbfz-diaaa-aaaaq-aacyq-cai", "Yuku");
+
+#[test]
+fn should_upgrade_icrc_ck_canisters_with_golden_state() {
+    let ledger_wasm = ledger_wasm();
+    let ledger_wasm_u256 = ledger_wasm_u256();
+
+    let canisters = vec![CK_BTC_LEDGER];
+    let canisters_u256 = vec![
+        CK_ETH_LEDGER,
+        CK_USDC_LEDGER,
+        CK_LINK_LEDGER,
+        CK_OCT_LEDGER,
+        CK_PEPE_LEDGER,
+        CK_SHIB_LEDGER,
+    ];
+
+    let state_machine = new_state_machine_with_golden_fiduciary_state_or_panic();
+
+    for canister in canisters {
+        upgrade_canister(&state_machine, canister, ledger_wasm.clone());
+    }
+    for canister_u256 in canisters_u256 {
+        upgrade_canister(&state_machine, canister_u256, ledger_wasm_u256.clone());
+    }
+}
+
+#[test]
+fn should_upgrade_icrc_sns_canisters_with_golden_state() {
+    let ledger_wasm = ledger_wasm();
+
+    let canisters = vec![
+        DRAGGINZ,
+        OPENCHAT,
+        ICPSWAP,
+        BOOMDAO,
+        CATALYZE,
+        ELNAAI,
+        ESTATEDAO,
+        GOLDDAO,
+        HOTORNOT,
+        ICGHOST,
+        ICLIGHTHOUSE,
+        ICPANDA,
+        ICPCC,
+        ICX,
+        KINIC,
+        DECIDEAI,
+        MOTOKO,
+        NEUTRINITE,
+        NUANCE,
+        OPENFPL,
+        ORIGYN,
+        SNEED,
+        SONIC,
+        TRAX,
+        WATERNEURON,
+        YUKU,
+    ];
+
+    let state_machine = new_state_machine_with_golden_sns_state_or_panic();
+
+    for canister in canisters {
+        upgrade_canister(&state_machine, canister, ledger_wasm.clone());
+    }
+}
+
+fn upgrade_canister(
+    state_machine: &StateMachine,
+    (canister_id_str, canister_name): (&str, &str),
+    ledger_wasm: Vec<u8>,
+) {
+    let canister_id =
+        CanisterId::unchecked_from_principal(PrincipalId::from_str(canister_id_str).unwrap());
+    upgrade_ledger(state_machine, ledger_wasm, canister_id);
```

Comment:
So we have been doing 2x upgrades in our tests to ensure that the first upgrade does not end up in a state where a second upgrade is not possible (i.e. pre_upgrade hook failure)

I don't know if that's covered elsewhere (haven't finished my review)

Reply:
I didn't find a place where this is done.  There should be examples in our tests, but we modify the wasm metadata so we can get a different hash, and then we do a second upgrade

Reply:
You're probably referring to [`bump_gzip_timestamp`](https://sourcegraph.com/github.com/dfinity/ic@fe231b38582fd12bb3e238e4e689fafa59b2b14f/-/blob/rs/nns/test_utils/src/itest_helpers.rs?L728-739). I'd heard about it, but I hadn't realized before that a 2x upgrade would help to catch a `pre_upgrade` hook failure - that's a great suggestion, and I'll add it to this test, and consider adding it to our other upgrade tests also!

Reply:
I added a second upgrade, but that second upgrade was then returning a `CanisterInstallCodeRateLimited` error. I asked about this [on Slack](https://dfinity.slack.com/archives/CGZJ7G1J6/p1721300686300469?thread_ts=1717412583.592399&cid=CGZJ7G1J6), and implemented the solution of disabling `rate_limiting_of_instructions`.

Reply:
Great, glad it was useful!